File: SunMoonChartView.m
//
//  FlatChartView.m
//  ShralpTide2
//
//  Created by Michael Parlee on 10/19/13.
//
//

#import "SunMoonChartView.h"
#import "NSDate+Day.h"
#import "UIImage+Mask.h"

@interface SunMoonChartView ()

@property (nonatomic,strong) NSDateFormatter *hourFormatter;
@property (nonatomic,strong) NSDateFormatter *timeFormatter;

@end

@implementation SunMoonChartView

- (id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self commonInit];
    }
    return self;
}

- (id)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self commonInit];
    }
    return self;
}

-(void)commonInit
{
    _labelInset = 0;
    _hourFormatter = [[NSDateFormatter alloc] init];
    _hourFormatter.dateFormat = [NSDateFormatter dateFormatFromTemplate:@"j" options:0 locale:[NSLocale currentLocale]];
    _timeFormatter = [[NSDateFormatter alloc] init];
    [_timeFormatter setDateStyle:NSDateFormatterNoStyle];
    [_timeFormatter setTimeStyle:NSDateFormatterShortStyle];
    self.sunriseIcon = [[UIImage imageNamed:@"sunrise_trnspt"] maskImageWithColor:[UIColor yellowColor]];
    self.sunsetIcon = [[UIImage imageNamed:@"sunset_trnspt"] maskImageWithColor:[UIColor orangeColor]];
    self.moonriseIcon = [[UIImage imageNamed:@"moonrise_trnspt"] maskImageWithColor:[UIColor whiteColor]];
    self.moonsetIcon = [[UIImage imageNamed:@"moonset_trnspt"] maskImageWithColor:[UIColor whiteColor]];
    self.showZero = NO;
}

- (void)drawRect:(CGRect)rect
{
    [super drawRect:rect];
    
	CGContextRef context = UIGraphicsGetCurrentContext();
    
    NSArray *intervalsForDay = [super.tide intervalsFromDate:[self.datasource day] forHours:self.hoursToPlot];
    if ([intervalsForDay count] == 0) {
        return;
    }
    NSTimeInterval baseSeconds = [((SDTideInterval*)intervalsForDay[0]).time timeIntervalSince1970];
    
    // draws the sun/moon events
    float lastX = 0;
    for (SDTideEvent *event in super.tide.sunAndMoonEvents) {
        UIImage *image = nil;
        if (event.eventType == sunrise) {
            image = self.sunriseIcon;
        } else if (event.eventType == sunset) {
            image = self.sunsetIcon;
        } else if (event.eventType == moonrise) {
            image = self.moonriseIcon;
        } else if (event.eventType == moonset) {
            image = self.moonsetIcon;
        }
        int minute = ([event.eventTime timeIntervalSince1970] - baseSeconds) / SECONDS_PER_MINUTE;
        float x = minute * self.xratio;
        CGSize size = CGSizeMake(15, 15);
        UIFont *font = [UIFont systemFontOfSize:13 weight:UIFontWeightBold];
        CGFloat imageHeight = _labelInset + font.lineHeight;
        CGRect rect = CGRectMake(x - size.width / 2, imageHeight, size.width, size.height);
        
        // draw the sunmoon times
        CGContextSetRGBStrokeColor(context, 1, 1, 1, 1);
        CGContextSetRGBFillColor(context, 1, 1, 1, 1);
        NSString *time = [[_timeFormatter stringFromDate:event.eventTime] stringByReplacingOccurrencesOfString:@" " withString:@""];
        
        CGFloat yOffset = 0;
        CGFloat sunHeight = _labelInset;
        CGFloat moonHeight = imageHeight + rect.origin.y;
        if ([@[@"Sunrise", @"Sunset"] containsObject:event.eventTypeDescription]) {
            yOffset = sunHeight;
        } else {
            yOffset = moonHeight;
        }
        CGFloat timeWidth = [time sizeWithAttributes: font.fontDescriptor.fontAttributes].width;
        CGFloat timeCenterX = x - timeWidth / 2;
        if (timeCenterX < 0) {
            timeCenterX = 0;
        } else if (x + (timeWidth / 2) >= self.frame.size.width) {
            timeCenterX = self.frame.size.width - timeWidth - 10;
        }
        [time drawAtPoint:CGPointMake(timeCenterX, yOffset) withAttributes:@{ NSFontAttributeName:font, NSForegroundColorAttributeName: [UIColor whiteColor] }];
        
        [image drawInRect:rect];
    }
}


@end

File: Classes/Views/SunMoonChartView.swift
import Foundation
import UIKit
import QuartzCore
import CoreGraphics


@objc class SunMoonChartView: ChartView {
    
    @objc var labelInset: CGFloat = 0
    
    fileprivate let hourFormatter = DateFormatter()
    fileprivate let timeFormatter = DateFormatter()
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        commonInit()
    }
    
    fileprivate func commonInit() {
        hourFormatter.dateFormat = "j"
        timeFormatter.dateStyle = .none
        timeFormatter.timeStyle = .short
        sunriseIcon = UIImage(named: "sunrise_trnspt")?.maskImage(with: .yellow)
        sunsetIcon = UIImage(named: "sunset_trnspt")?.maskImage(with: .orange)
        moonriseIcon = UIImage(named: "moonrise_trnspt")?.maskImage(with: .white)
        moonsetIcon = UIImage(named: "moonset_trnspt")?.maskImage(with: .white)
    }
    
    override func draw(_ rect: CGRect) {
        super.draw(rect)
        
        guard let tide = datasource.tideDataToChart else {
            return
        }
        
        let context = UIGraphicsGetCurrentContext()!
        
        let intervalsForDay:[SDTideInterval] = tide.intervals(from: datasource.day, forHours: hoursToPlot)
        if intervalsForDay.count == 0 { return }
        let baseSeconds = intervalsForDay[0].time.timeIntervalSince1970
        tide.sunAndMoonEvents.forEach { event in
            do {
                let image = try self.image(forEvent: event)
                let minute = Int(event.eventTime.timeIntervalSince1970 - baseSeconds) / ChartView.SecondsPerMinute
                let x = CGFloat(minute) * self.xratio
                let size = CGSize(width: 15, height: 15)
                let font = UIFont.systemFont(ofSize: 13, weight: .bold)
                let imageHeight = labelInset + font.lineHeight;
                let rect = CGRect(x: x - size.width / 2, y: imageHeight, width: size.width, height: size.height)
                
                context.setStrokeColor(red: 1, green: 1, blue: 1, alpha: 1)
                context.setFillColor(red: 1, green: 1, blue: 1, alpha: 1)
                
                let time = timeFormatter.string(from: event.eventTime).replacingOccurrences(of: " ", with: "")
                var yOffset:CGFloat = 0
                let sunHeight: CGFloat = labelInset
                let moonHeight: CGFloat = imageHeight + rect.origin.y
                if ["Sunrise", "Sunset"].contains(event.eventTypeDescription) {
                    yOffset = sunHeight
                } else {
                    yOffset = moonHeight
                }
                let attributes = [NSAttributedString.Key.font: font]
                let timeWidth:CGFloat = time.size(withAttributes: attributes).width
                var timeCenterX = x - timeWidth / 2
                if timeCenterX < 0 {
                    timeCenterX = 0
                } else if x + (timeWidth / 2) + 10 >= frame.size.width {
                    timeCenterX = frame.size.width - timeWidth - 10
                }
                time.draw(at: CGPoint(x: timeCenterX, y: yOffset),
                          withAttributes: [ .font: font,
                                            .foregroundColor: UIColor.white
                    ]
                )
                image.draw(in: rect)
                
                context.strokePath()
            } catch {
                print("Unexpected error: \(error)")
            }
        }
    }
    
    fileprivate func image(forEvent event: SDTideEvent) throws -> UIImage {
        switch (event.eventType) {
        case .sunrise:
            return sunriseIcon!
        case .sunset:
            return sunsetIcon!
        case .moonrise:
            return moonriseIcon!
        case .moonset:
            return moonsetIcon!
        default:
            throw ChartError.eventNotSupported
        }
    }
}
